### PR TITLE
Fix block hash

### PR DESCRIPTION
**Pull request by zjg555543**
To fix xlayer mainnet. https://github.com/0xPolygonHermez/cdk-erigon/issues/576

### CODE REVIEW

File: core/vm/instructions_zkevm.go [hunk 1]
```diff
@@ -225,14 +225,6 @@ func makeLog_zkevm(size int, logIndexPerTx bool) executionFunc {
 				}
 
 				d, _ = hex.DecodeString(dataHex)
-			} else {
-				// erigon behaviour
```
Comment (revitteth) [inline core/vm/instructions_zkevm.go:229]: Changeset looks good, however I don't think we can remove this - there is a case where the bug is not possible, above the 'bug trigger blockheight' where we will not correctly populate 'd'?

Comment (zjg555543) [reply core/vm/instructions_zkevm.go:229]: But in this case, the log data does not need padding zeros in forkId.9. So I am a little confused
```
# This is a zkevm node rpc result with forkID.9
curl -X POST -H "Content-Type: application/json" -d '{"jsonrpc":"2.0","method":"eth_getTransactionReceipt","params":["0x6118d61806013e84408a47bd1bad1e44149534bc2f4b4dae5f440fe11618c579"],"id":1}' https://rpc.xlayer.tech | jq
```

Comment (mandrigin) [reply core/vm/instructions_zkevm.go:229]: What about zkevm mainnet?

Comment (zjg555543) [reply core/vm/instructions_zkevm.go:229]: Got it, I will test zkevm mainnet

**Review by mandrigin — CHANGES_REQUESTED** [review]
We need to check with Carlos about the exact behavior — currently unclear when and how it should be activated. But this extra code definitely should not be removed.

**Comment by mandrigin** [comment]
@zjg555543 when was the network upgraded and to which forkIDs? Can you give us a table for XLayer?

Like 
* forkid.6 — genesis to block N
* forkid.8 — block N+1 to M
* forkid.9 — block M+1 to now

File: core/vm/instructions_zkevm.go (hunk 1 above)
Comment (zjg555543) [reply core/vm/instructions_zkevm.go:229]: I have tested on zkevm mainnet with branch **okx:zjg/fix-block-hash**, It got the same block hash form zkevm node.
So I think we can delete [zkEvm] fill 0 at the end?
```
# erigon zjg/fix-block-hash rpc node
 curl -X POST -H "Content-Type: application/json" -d '{"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["13105691",true],"id":1}' http://127.0.0.1:8545/
```
```
{"jsonrpc":"2.0","id":1,"result":{"difficulty":"0x0","extraData":"0x","gasLimit":"0x4000000000000","gasUsed":"0x5208","hash":"0x68a36749226cfcb9e50611c2069921b5372e3559d80d34aba5208d4ce4333dae","logsBloom":"0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000","miner":"0x148ee7daf16574cd020afa34cc658f8f3fbd2800","mixHash":"0x0000000000000000000000000000000000000000000000000000000000000000","nonce":"0x0000000000000000","number":"0xc7fa1b","parentHash":"0x3be47dbf1ac29ffae5822801ed3bc0068995360bf6d62832f42b23017e597fa7","receiptsRoot":"0x056b23fbba480696b65fe5a59b8f2148a1299103c4f57df839233af2cf4ca2d2","sha3Uncles":"0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347","size":"0x276","stateRoot":"0x0dee161078a0adfe25fce3d398f20f7f00bba92585d9fe621db003003f55c488","timestamp":"0x6660643e","totalDifficulty":"0x0","transactions":[{"blockHash":"0x68a36749226cfcb9e50611c2069921b5372e3559d80d34aba5208d4ce4333dae","blockNumber":"0xc7fa1b","from":"0x9af3049dd15616fd627a35563b5282bea5c32e20","gas":"0x5208","gasPrice":"0x472afe80","hash":"0x9e3d21eb0d1636bad495ca235b827882fa6b7c2bc4c4d7ff37e23e9be5b4da39","input":"0x","nonce":"0x70870","to":"0x417a7ba2d8d0060ae6c54fd098590db854b9c1d5","transactionIndex":"0x0","value":"0x9184e72a000","type":"0x0","chainId":"0x44d","v":"0x8bd","r":"0xeb935366231fc74c3c5289bac8c78e7987c1d6c64e2b66ba64795bba893cca7f","s":"0x7b4eecea7d7ba472f45a6a65b28c2173f9eab74466ae23bcd57a06e1192a430e"}],"transactionsRoot":"0x6fc12a5a2b9c9668291270175d56df0d61d902a088c8b2e9c0ae154bdde8c092","uncles":[]}}
```

**Comment by zjg555543** [comment]
> @zjg555543 when was the network upgraded and to which forkIDs? Can you give us a table for XLayer?
> 
> Like
> 
> * forkid.6 — genesis to block N
> * forkid.8 — block N+1 to M
> * forkid.9 — block M+1 to now

@mandrigin The issue only occurs on the XLayer mainnet, which has forkID.9 and has not been upgraded.
https://www.oklink.com/zh-hans/xlayer/tx/0x6118d61806013e84408a47bd1bad1e44149534bc2f4b4dae5f440fe11618c579/log

Here is the mainnet table.
```
state_db=# select * from state.fork_id;
 fork_id | from_batch_num |     to_batch_num     | version | block_num 
---------+----------------+----------------------+---------+-----------
       9 |              1 | 18446744073709551615 |         |       374
```

**Comment by zjg555543** [comment]
BTW, we can ignore the mismatch issues for the XLayer Testnet. 

We plan to forcefully modify the latest block hash of the sequencer and the permissionless RPC to match Erigon, and it will ignore any mismatch blocks in between.

Here is the testnet's table.
```
state_db=# select * from state.fork_id;
 fork_id | from_batch_num |     to_batch_num     |     version     | block_num 
---------+----------------+----------------------+-----------------+-----------
       5 |              1 |                49038 | 0.1.0           |   4648290
       6 |          49039 |               321733 | x1-fork6-v0.2.0 |   4915702
       8 |         321734 |               476000 |                 |   5482576
       9 |         476001 | 18446744073709551615 |                 |   5681756
```

File: core/vm/instructions_zkevm.go (hunk 1 above)
Comment (mandrigin) [reply core/vm/instructions_zkevm.go:229]: Cool, we probably need to make sure that it also works on Bali & Cardona before we can merge…

Comment (mandrigin) [reply core/vm/instructions_zkevm.go:229]: But overall looks all right to me.

**Comment by mandrigin** [comment]
@zjg555543 we intend to keep full compatibility on block hashes for all forkIDs for now, until Erigon3 migration is done. So if you have block hash mismatches on the testnet — please report them.

**Comment by mandrigin** [comment]
@zjg555543 can you check if this change set instead solves your issue? https://github.com/0xPolygonHermez/cdk-erigon/pull/565

File: core/vm/instructions_zkevm.go (hunk 1 above)
Comment (zjg555543) [reply core/vm/instructions_zkevm.go:229]: > Cool, we probably need to make sure that it also works on Bali & Cardona before we can merge…

Okay, I will test on Bali & Cardona. Maybe Bali use protobuffer ds, I need create an new branch first to test on Bali.

**Comment by zjg555543** [comment]
> can you check if this change set instead solves your issue?

I don't think it solves the issue on xlayer mainnet which use forkID.9. For this tx case, i didn't fil 0 for log data which len is 712.
image

``` shell
# This is a zkevm node rpc result with forkID.9
curl -X POST -H "Content-Type: application/json" -d '{"jsonrpc":"2.0","method":"eth_getTransactionReceipt","params":["0x6118d61806013e84408a47bd1bad1e44149534bc2f4b4dae5f440fe11618c579"],"id":1}' https://rpc.xlayer.tech | jq
```

**Comment by zjg555543** [comment]
> @zjg555543 when was the network upgraded and to which forkIDs? Can you give us a table for XLayer?
> 
> Like
> 
> * forkid.6 — genesis to block N
> * forkid.8 — block N+1 to M
> * forkid.9 — block M+1 to now


Thank you for your support. We have discussed internally and decided not to attempt fixing this issue on the xlayer testnet. This issue might be caused by our upgrade fork6~fork8, but it is not important as long as the blocks after forkID9 are consistent. We are planning to enforce a rehash on seq and perssionless rpc soon to ensure consistency with erigon. If a mismatch occurs again after that, we will need to pay special attention.

**Comment by mandrigin** [comment]
Can you send us the tx hashes where the current Erigon has wrong logs and we will review with Carlos and Max? It feels like we miss something there and I don’t really want to include a fix I don’t understand exactly.

File: core/vm/instructions_zkevm.go (hunk 1 above)
Comment (zjg555543) [reply core/vm/instructions_zkevm.go:229]: - Cardona tests successed.Eriong  and zkEVM rpc 's latest block hash is same.

- But, Bali has an issue like this. both the zkevm or my delete filling 0 data branch have this issue.
```
INFO] [06-06|14:04:10.807] [9/16 IntermediateHashes] Finished the temp tree build in 1m48.988903995s, hashing and saving the result...
[INFO] [06-06|14:04:10.808] [9/16 IntermediateHashes] Regeneration ended
[INFO] [06-06|14:04:10.808] [9/16 IntermediateHashes] Trie root      hash=0x0000000000000000000000000000000000000000000000000000000000000000
panic: [9/16 IntermediateHashes] Wrong trie root of block 2441704: 0000000000000000000000000000000000000000000000000000000000000000, expected (from header): 5576df1e1b201499c87fe44ac61735973a44aa854d5ed0ce564639c76cba1e40. Block hash: d97bb839d46475f04cbaec76ee707ce9d4affb29ce53b217d4c9003969a60238

goroutine 53 [running]:
github.com/ledgerwatch/erigon/zk/stages.SpawnZkIntermediateHashesStage(0xc00b1d3d00, {0x2eaa598?, 0xc000236020?}, {0x0, 0x0}, {{0x2ebcc78, 0xc0012047e0}, 0x1, 0x0, {0xc000a032c0, ...}, ...}, ...)
        github.com/ledgerwatch/erigon/zk/stages/stage_interhashes.go:171 +0x1234
github.com/ledgerwatch/erigon/zk/stages.DefaultZkStages.func25(0x40?, 0xc8?, 0
```

**Comment by zjg555543** [comment]
> Can you send us the tx hashes where the current Erigon has wrong logs and we will review with Carlos and Max? It feels like we miss something there and I don’t really want to include a fix I don’t understand exactly.

@mandrigin Thank you for your attention.  I log an issue to describe the mismatch issue. https://github.com/0xPolygonHermez/cdk-erigon/issues/576

**Comment by mandrigin** [comment]
fixed by @revitteth in #595

**Pull request closed without merging**